Content.Server/Explosion/EntitySystems/ProjectileGrenadeSystem.cs

```diff
@@ -0,0 +1,111 @@
+
+public sealed class ProjectileGrenadeSystem : EntitySystem
+{
+    [Dependency] private readonly GunSystem _gun = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedContainerSystem _container = default!;
+    [Dependency] private readonly TransformSystem _transformSystem = default!;
+
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ProjectileGrenadeComponent, ComponentInit>(OnFragInit);
+        SubscribeLocalEvent<ProjectileGrenadeComponent, ComponentStartup>(OnFragStartup);
+        SubscribeLocalEvent<ProjectileGrenadeComponent, TriggerEvent>(OnFragTrigger);
+    }
+
+    private void OnFragInit(Entity<ProjectileGrenadeComponent> entity, ref ComponentInit args)
+    {
+        entity.Comp.Container = _container.EnsureContainer<Container>(entity.Owner, "cluster-payload");
+    }
+
+    /// <summary>
+    /// Setting the unspawned count based on capacity so we know how many new entities to spawn
+    /// </summary>
+    private void OnFragStartup(Entity<ProjectileGrenadeComponent> entity, ref ComponentStartup args)
+    {
+        if (entity.Comp.FillPrototype == null)
+            return;
+
+        entity.Comp.UnspawnedCount = Math.Max(0, entity.Comp.Capacity - entity.Comp.Container.ContainedEntities.Count);
+    }
+
+    /// <summary>
+    /// Can be triggered either by damage or the use in hand timer
+    /// </summary>
+    private void OnFragTrigger(Entity<ProjectileGrenadeComponent> entity, ref TriggerEvent args)
+    {
+        FragmentIntoProjectiles(entity.Owner, entity.Comp);
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Spawns projectiles at the coordinates of the grenade upon triggering
+    /// Can customize the angle and velocity the projectiles come out at
+    /// </summary>
+    private void FragmentIntoProjectiles(EntityUid uid, ProjectileGrenadeComponent component)
+    {
+        var grenadeCoord = _transformSystem.GetMapCoordinates(uid);
+        var shootCount = 0;
+        var totalCount = component.Container.ContainedEntities.Count + component.UnspawnedCount;
+        var segmentAngle = 360 / totalCount;
+
+        while (TrySpawnContents(grenadeCoord, component, out var contentUid))
+        {
+            Angle angle;
+            if (component.RandomAngle)
+                angle = _random.NextAngle();
+            else
+            {
+                var angleMin = segmentAngle * shootCount;
+                var angleMax = segmentAngle * (shootCount + 1);
+                angle = Angle.FromDegrees(_random.Next(angleMin, angleMax));
+                shootCount++;
+            }
+
+            // velocity is randomized to make the projectiles look
+            // slightly even, doesn't really change much but it looks better
+            Vector2 direction = angle.ToVec().Normalized();
+            Vector2 velocity = _random.NextVector2(component.Velocity);
```

Comment:
So this can generate a value between 0 and component.Velocity? This should have a minimum value

Reply:
`public void ShootProjectile(EntityUid uid, Vector2 direction, Vector2 gunVelocity, EntityUid gunUid, EntityUid? user = null, float speed = 20f)
    {
        var physics = EnsureComp<PhysicsComponent>(uid);
        Physics.SetBodyStatus(uid, physics, BodyStatus.InAir);

        var targetMapVelocity = gunVelocity + direction.Normalized() * speed;`

even if the provided velocity ends up being 0, apparently the ShootProjectile function has a base speed of 20f plus whatever additional velocity is provided

this does essentially make the component Velocity completely useless... I guess, but it doesn't cause any problems.